File: flock/c/flock.c

#include <stdio.h>
#include <stdlib.h>
#include <sys/file.h>

int main(int argc, char* argv[]) {

	char *line;
	FILE *fd;
	ssize_t res;
	size_t n;
	int fres;

	if (2 != argc) {
		fprintf(stderr, "usage: %s <file>\n", argv[0]);
		return 1;
	}

	// open, create, truncate
	fd = fopen(argv[1], "w+");
	if (NULL == fd) {
		perror("fopen");
		return 1;
	}

	printf("acquiring lock...");
	fflush(stdout);
	fres = flock(fileno(fd), LOCK_EX);
	if (-1 == fres) {
		perror("flock");
		return 1;
	}
	printf("done\n");

	printf("Enter data: ");

	line = NULL;
	n = 0;
	res = getline(&line, &n, stdin);
	if (-1 == res) {
		perror("getline");
		return 1;
	}

	res = fwrite(line, n, 1, fd);

	free(line);

	printf("releasing lock...");
	fres = flock(fileno(fd), LOCK_UN);
	if (-1 == fres) {
		perror("flock");
		return 1;
	}
	printf("done\n");

	fclose(fd);

	return 0;
}

File: flock/python/flock.py
#!/usr/bin/python

import fcntl
import sys

if (len(sys.argv) != 2):
    sys.stderr.write('usage: ' + sys.argv[0] + ' <file>\n')
    sys.exit(1)

fd = open(sys.argv[1], 'w+')

sys.stdout.write('acquiring lock...')
sys.stdout.flush()
fcntl.flock(fd.fileno(), fcntl.LOCK_EX)
sys.stdout.write('done\n')

line = raw_input('Enter data: ')

fd.write(line)

sys.stdout.write('releasing lock...')
fcntl.flock(fd.fileno(), fcntl.LOCK_UN)
sys.stdout.write('done\n')

fd.close()

sys.exit(0)
